controllers/nutanixcluster_controller.go

```diff
@@ -289,20 +321,94 @@ func (r *NutanixClusterReconciler) reconcileNormal(rctx *nctx.ClusterContext) (r
+
+	// First clear the failure domain related conditions
+	conditions.Delete(rctx.NutanixCluster, infrav1.FailureDomainsValidatedCondition)
+	conditions.Delete(rctx.NutanixCluster, infrav1.NoFailureDomainsConfiguredCondition)
+
+	failureDomains := capiv1.FailureDomains{}
+
+	if len(rctx.NutanixCluster.Spec.ControlPlaneFailureDomains) == 0 && len(rctx.NutanixCluster.Spec.FailureDomains) == 0 {
+		log.Info("No failure domains configured for cluster.")
+		conditions.MarkTrue(rctx.NutanixCluster, infrav1.NoFailureDomainsConfiguredCondition)
+		// Reset the failure domains for nutanixcluster status
+		rctx.NutanixCluster.Status.FailureDomains = failureDomains
 		return nil
 	}
-	log.V(1).Info("Reconciling failure domains for cluster")
-	// If failure domains is nil on status object, first create empty slice
-	if rctx.NutanixCluster.Status.FailureDomains == nil {
-		rctx.NutanixCluster.Status.FailureDomains = make(capiv1.FailureDomains, 0)
+
+	// Not found failure domain names
+	notFoundFDs := []string{}
+	// Map of mis-configured failure domain name and validation error message
+	misConfigFDs := map[string]string{}
+
+	for _, fdRef := range rctx.NutanixCluster.Spec.ControlPlaneFailureDomains {
+		// Fetch the referent failure domain object
+		fdObj := &infrav1.NutanixFailureDomainV2{}
+		fdKey := client.ObjectKey{Name: fdRef.Name, Namespace: rctx.NutanixCluster.Namespace}
+		err := r.Client.Get(rctx.Context, fdKey, fdObj)
+		if err != nil {
+			log.Error(err, fmt.Sprintf("Failed to fetch the referent failure domain object %q.", fdRef.Name))
+			notFoundFDs = append(notFoundFDs, fdRef.Name)
+			continue
+		}
+
+		// Validate the failure domain configuration
+		err = r.validateFailureDomainSpec(rctx, fdObj)
+		if err != nil {
+			log.Error(err, fmt.Sprintf("The failure domain %q configuration failed with validation.", fdObj.Name))
+			misConfigFDs[fdObj.Name] = err.Error()
+			continue
+		}
+
+		failureDomains[fdObj.Name] = capiv1.FailureDomainSpec{ControlPlane: true}
 	}
+
+	// Remove below when the Deprecated field NutanixCluster.Spec.FailureDomains is removed
 	for _, fd := range rctx.NutanixCluster.Spec.FailureDomains {
-		rctx.NutanixCluster.Status.FailureDomains[fd.Name] = capiv1.FailureDomainSpec{ControlPlane: fd.ControlPlane}
+		failureDomains[fd.Name] = capiv1.FailureDomainSpec{ControlPlane: fd.ControlPlane}
+	}
+
+	// Set the failure domains for nutanixcluster status
+	rctx.NutanixCluster.Status.FailureDomains = failureDomains
+
+	if len(notFoundFDs) == 0 && len(misConfigFDs) == 0 {
+		conditions.MarkTrue(rctx.NutanixCluster, infrav1.FailureDomainsValidatedCondition)
+	} else {
+		errMsg := ""
+		if len(notFoundFDs) > 0 {
+			errMsg = fmt.Sprintf("Not-found failure domain objects: %v", notFoundFDs)
+		}
+		if len(misConfigFDs) > 0 {
+			errMsg1 := fmt.Sprintf("Mis-configured failure domains: %v", misConfigFDs)
+			if errMsg == "" {
+				errMsg = errMsg1
+			} else {
+				errMsg += ", " + errMsg1
+			}
+		}
```

Comment:
this code is confusing to read. please refactor it and please use more descriptive variables

Reply:
Add more comment to explain